File: bot/src/quotes/curve.ts
import { PublicClient, Address, encodeFunctionData } from "viem";

const CURVE_ABI = [
  {
    name: "get_dy",
    type: "function",
    stateMutability: "view",
    inputs: [
      { name: "i", type: "int128" },
      { name: "j", type: "int128" },
      { name: "dx", type: "uint256" }
    ],
    outputs: [
      { name: "dy", type: "uint256" }
    ]
  }
] as const;

export async function quoteCurveGetDy(
  client: PublicClient, 
  pool: Address, 
  i: bigint, 
  j: bigint, 
  dx: bigint
): Promise<bigint> {
  try {
    const data = encodeFunctionData({
      abi: CURVE_ABI,
      functionName: "get_dy",
      args: [i, j, dx]
    });

    const result = await client.call({
      to: pool,
      data
    });

    if (!result.data) {
      throw new Error("No data returned from Curve pool");
    }

    // Decode the result (dy is the first return value)
    const dy = BigInt(result.data.slice(2, 66)); // First 32 bytes
    return dy;
  } catch (error) {
    console.error("Error quoting Curve pool:", error);
    return 0n;
  }
}

File: dex_arb/pybot/quotes/curve.py
from web3 import Web3
from typing import Union

# Curve get_dy ABI
GET_DY_ABI = [{
    "inputs": [
        {"internalType": "int128", "name": "i", "type": "int128"},
        {"internalType": "int128", "name": "j", "type": "int128"},
        {"internalType": "uint256", "name": "dx", "type": "uint256"}
    ],
    "name": "get_dy",
    "outputs": [{"internalType": "uint256", "name": "dy", "type": "uint256"}],
    "stateMutability": "view",
    "type": "function"
}]

# Curve get_dy_underlying ABI (for meta pools)
GET_DY_UNDERLYING_ABI = [{
    "inputs": [
        {"internalType": "int128", "name": "i", "type": "int128"},
        {"internalType": "int128", "name": "j", "type": "int128"},
        {"internalType": "uint256", "name": "dx", "type": "uint256"}
    ],
    "name": "get_dy_underlying",
    "outputs": [{"internalType": "uint256", "name": "dy", "type": "uint256"}],
    "stateMutability": "view",
    "type": "function"
}]

def get_dy(w3: Web3, pool: str, i: int, j: int, dx: int, underlying: bool = False) -> int:
    """Get Curve pool quote for token swap."""
    contract = w3.eth.contract(
        address=Web3.to_checksum_address(pool), 
        abi=GET_DY_UNDERLYING_ABI if underlying else GET_DY_ABI
    )
    
    if underlying:
        return contract.functions.get_dy_underlying(i, j, dx).call()
    else:
        return contract.functions.get_dy(i, j, dx).call()
